### PR TITLE
remove dummy data generated to just sneak past validation

### PR DESCRIPTION
**Pull request by KeynesYouDigIt**
Note that its just a draft, can @lognaturel or @MartijnR  confirm that the tests conform to what #286 is asking for? That is, if these tests pass286 is resolved?

### CODE REVIEW

**Comment by lognaturel** [comment]
Thanks for taking this on and for your patience, @KeynesYouDigIt! @MartijnR is correct that there is a Validate issue. I thought updating to a JavaRosa version that supported external secondary itemsets was enough but it's not. I have a pull request at https://github.com/opendatakit/validate/pull/66 that should address the issue.

Your test changes look good to me. One small thing I think I mentioned, perhaps in Slack -- if `instance` could be a self-closing tag I would like that better.

**Comment by lognaturel** [comment]
https://github.com/XLSForm/pyxform/pull/320 updates Validate to the latest which will successfully validate forms with XML external instances and no dummy data. However, it doesn't yet work for CSV ones (blocked on https://github.com/opendatakit/javarosa/issues/417). I suggest we wait until that is fixed to merge this PR (sorry).

In the mean time, I took a closer look at the tests and @KeynesYouDigIt, I don't think you need a new test for this! You'll need to modify the existing ones anyway (as it looks like you've already done) since the expected output has changed and I think that's verification enough that the dummy nodes aren't generated.

> if instance could be a self-closing tag I would like that better.

The actual output does include self-closing tags so I think that will need to change. What you could do now, @KeynesYouDigIt, is get the tests passing with `run_odk_validate=False` and then all that would be needed is to flip the switch once the CSV update is available.

**Comment by KeynesYouDigIt** [comment]
Looks like https://github.com/opendatakit/javarosa/issues/417 was resolved by https://github.com/opendatakit/javarosa/pull/452 so Im going to jump back in, do I have to worry about my version of Java Rosa? still getting some validate errors like 

```
ODK Validate error was thrown but 'odk_validate_error__contains'
was empty:b"ODK Validate Errors:\n>> XForm is invalid.
See above for the errors.\norg.javarosa.xform.parse.XFormParseException:
<label> node for itemset doesn't exist! [instance(cities)/root/item/label]
\n\nThe following files failed validation:\n${tmpy6nqd6j8}.xml\n\nResult: Invalid"
```


But I believe I can figure them out, hints are welcome if anyone is around/not caught up in something

**Comment by KeynesYouDigIt** [comment]
APU, no rush on this as I will tinker around to figure it out.

So, with these changes heres the diff in the XML. 

with my changes
```xml
        <model>
            <instance>
                <pyxform_autotestname id="pyxform_autotest_id_string">
                    <city/>
                    <neighbourhoods/>
                    <meta>
                        <instanceID/>
                    </meta>
                </pyxform_autotestname>
            </instance>
            <instance id="cities" src="jr://file-csv/cities.csv"/>
            <instance id="neighbourhoods" src="jr://file-csv/neighbourhoods.csv"/>
```

master
```xml
        <model>
            <instance>
                <pyxform_autotestname id="pyxform_autotest_id_string">
                    <city/>
                    <neighbourhoods/>
                    <meta>
                        <instanceID/>
                    </meta>
                </pyxform_autotestname>
            </instance>
            <instance id="cities" src="jr://file-csv/cities.csv">
                <root>
                    <item>
                        <name>_</name>
                        <label>_</label>
                    </item>
                </root>
            </instance>
            <instance id="neighbourhoods" src="jr://file-csv/neighbourhoods.csv">
                <root>
                    <item>
                        <name>_</name>
                        <label>_</label>
                    </item>
                </root>
            </instance>
```


If I understand the original issue, my code does whats needed, but the validator is complaining about not having items (see previous comment).

Do I need to update my validator somehow? cant find in the readme where it talks about installing validate.

**Comment by lognaturel** [comment]
@yanokwa has since noticed a couple more issues with Validate. 😫 See https://github.com/opendatakit/validate/releases for the latest release that addresses them.

You should be able to download that jar, rename it to `ODK_Validate.jar` and put it [here](https://github.com/XLSForm/pyxform/tree/master/pyxform/validators/odk_validate/bin) to try out your changes. In fact, I think it would be appropriate to do the update in this PR. Please put the update in its own commit so it's easy to move elsewhere if @ukanga would prefer that change came separately.

**Comment by yanokwa** [comment]
@KeynesYouDigIt You can try rebasing on master and see if that works. If not, the easiest way to update Validate is to run `pyxform_validator_update odk update ODK-Validate-v1.13.1.jar` from your command line.

**Comment by KeynesYouDigIt** [comment]
Thanks @yanokwa and @lognaturel !! morbid curiosity - why is my xform valid? it looks like it would produce a question with no choices or input?

File: pyxform/validators/odk_validate/latest.json [hunk 1]
```diff
@@ -0,0 +1,74 @@
+{
+  "assets": [
```
Comment (KeynesYouDigIt) [inline pyxform/validators/odk_validate/latest.json]: Do we need this file? I think the update command auto generated it?

File: pyxform/survey.py [hunk 2]
```diff
@@ -231,11 +231,6 @@ def _generate_static_instances(list_name, choice_list):
             ),
         )
 
-    @staticmethod
-    def _get_dummy_instance():
-        """Instance content required by ODK Validate for select inputs."""
-        return node("root", node("item", node("name", "_"), node("label", "_")))
-
```
Comment (KeynesYouDigIt) [inline pyxform/survey.py]: This is really the substance of this PR, everything else adjusts around this change

**Comment by MartijnR** [comment]
> Does Enketo need to know about this change?

It knows. Thanks!

**Comment by MartijnR** [comment]
@KeynesYouDigIt fyi, [this](https://enketo.org/xforms/#csv-support) is the transformation that Enketo/ODK Collect makes to turn CSV into XML. This is the missing info to make sense of the XForm (`<itemset>`) syntax. This is not yet documented in the ODK XForms spec but will be soo.

**Comment by KeynesYouDigIt** [comment]
Please LMK if its mergeable, squashing now

**Comment by KeynesYouDigIt** [comment]
@ukanga :shipit: ?

**Comment by KeynesYouDigIt** [comment]
Anyone around for a look? @ukanga ?

**Comment by lognaturel** [comment]
I've verified that: 
- a resulting form looks as I would expect
- the tests pass
- if I add `file-csv` to Collect's root translators, [this form](https://docs.google.com/spreadsheets/d/1xZDS2XqwbMcgqa5nvi5_Ug03WkijogcQh4yPNexqaE8/edit?usp=sharing) works with both CSV and XML external files and both single and multiple selects

I can't think of any risk here now that all the pieces are playing nice together. The dummy nodes were required because there was no `from_file` support in Collect but that has now been addressed. Users of older versions of Collect would not have forms with `from_file` because it wouldn't work at all. @MartijnR has confirmed that Enketo is happy either way.

Quick retrospective on why this took so painfully long:
- I thought that updating JavaRosa with external secondary instance support was sufficient
- It was not because forms with external secondary instances could only be instantiated in the presence of those external data files
- When pyxform converts a form definition, it doesn't (and shouldn't) care about attachments
- I patched Validate in a way that I thought would work for both XML and CSV
- But then I found out the CSV support in JavaRosa was broken
- Yaw realized that the CSV support in Validate was incomplete as well

I think there are two root causes here:
- external secondary instance support has been unfortunately as secondary priority and is getting patched together on the ODK core tool side as folks/I get some free time. This has led to sometimes short-sighted solutions rather than diving deep to analyze problems and solve them once and for all. I think we're close on this one but for interventions of this scale, we should try harder to block off some time so issues in all dependent tools can be ironed out in one batch.
- We/I keep forgetting that Validate doesn't just parse the form but also runs through the form as Collect does. This allows it to catch classes of problems that are very valuable to users but also means it needs answers for things like where external files are (though in this case, it turns out parsing the form also requires the external file which is a separate problem). I've submitted https://github.com/opendatakit/validate/pull/78 to make it clearer.

If anyone else (especially @KeynesYouDigIt) has thoughts on how we can do better in the future, please share.

File: pyxform/tests_v1/test_support_external_instances.py [hunk 3]
```diff
@@ -7,7 +7,10 @@
 
 class ExternalCSVInstancesTest(PyxformTestCase):
     def test_external_csv_instances(self):
+
         # re: https://github.com/XLSForm/pyxform/issues/30
+
+        # this is probabbbbly the test that needs to pass before pushing
```
[the remark this answers is not in the document]
Comment (KeynesYouDigIt) [reply pyxform/tests_v1/test_support_external_instances.py]: 1a8eb673bc24ed81ee91a0f1816d428d0211396c should do it!

File: pyxform/tests_v1/test_support_external_instances.py [hunk 4]
```diff
@@ -7,7 +7,10 @@
 
 class ExternalCSVInstancesTest(PyxformTestCase):
     def test_external_csv_instances(self):
+
```
[the remark this answers is not in the document]
Comment (KeynesYouDigIt) [reply pyxform/tests_v1/test_support_external_instances.py]: 1a8eb673bc24ed81ee91a0f1816d428d0211396c should do it!

File: pyxform/tests_v1/test_xmldata.py [hunk 5]
```diff
@@ -448,10 +411,10 @@ def test_external_instances_multiple_diff_pulldatas(self):
...
         """
-        node1 = '<instance id="fruits" src="jr://file-csv/fruits.csv">'
-        node2 = '<instance id="OtherID" src="jr://file-csv/OtherID.csv">'
+        node1 = '<instance id="fruits" src="jr://file-csv/fruits.csv"/>'
+        node2 = '<instance id="OtherID" src="jr://file-csv/OtherID.csv"/>'
 
         survey = self.md_to_pyxform_survey(md_raw=md)
         xml = survey._to_pretty_xml()
-        self.assertEqual(1, xml.count(node1))
+        self.assertEqual(0, xml.count(node1))
```
[the remark this answers is not in the document]
Comment (KeynesYouDigIt) [reply pyxform/tests_v1/test_xmldata.py]: Not sure what I was thinking here, mustve assumped node 1 was a dummy node but its so clearly not. Fixing...

[the remark this answers is not in the document]
Comment (KeynesYouDigIt) [reply pyxform/tests_v1/test_xmldata.py]: !()[https://media.giphy.com/media/jPAdK8Nfzzwt2/giphy.gif]

File: pyxform/survey.py [hunk 6]
```diff
@@ -308,22 +301,14 @@ def get_pulldata_functions(element):
...
-                            name=file_id,
-                            src=uri,
-                            instance=node(
-                                "instance",
-                                Survey._get_dummy_instance(),
-                                id=file_id,
-                                src=uri,
-                            ),
-                        )
+                    return InstanceInfo(
```
Comment (KeynesYouDigIt) [inline pyxform/survey.py]: Why am I the way that I am? silliness here fixint

**Comment by KeynesYouDigIt** [comment]
@lognaturel Should be good for another look!

**Pull request merged**